hoodie-hive/src/main/java/com/uber/hoodie/hive/HoodieHiveClient.java

```diff
@@ -0,0 +1,607 @@
+        schema.put(columnName, columnType);
+      }
+      return schema;
+    } catch (SQLException e) {
+      throw new HoodieHiveSyncException(
+          "Failed to get table schema for " + syncConfig.tableName, e);
+    } finally {
+      closeQuietly(result, null);
+    }
+  }
+
+  /**
+   * Gets the schema for a hoodie dataset.
+   * Depending on the type of table, read from any file written in the latest commit.
+   * We will assume that the schema has not changed within a single atomic write.
+   *
+   * @return Parquet schema for this dataset
+   */
+  @SuppressWarnings("WeakerAccess")
+  public MessageType getDataSchema() {
+    try {
+      switch (tableType) {
+        case COPY_ON_WRITE:
+          // If this is COW, get the last commit and read the schema from a file written in the last commit
+          HoodieInstant lastCommit = activeTimeline.lastInstant()
+              .orElseThrow(() -> new InvalidDatasetException(syncConfig.basePath));
+          HoodieCommitMetadata commitMetadata = HoodieCommitMetadata
+              .fromBytes(activeTimeline.getInstantDetails(lastCommit).get());
+          String filePath = commitMetadata.getFileIdAndFullPaths().values().stream().findAny()
+              .orElseThrow(() -> new IllegalArgumentException(
+                  "Could not find any data file written for commit " + lastCommit
+                      + ", could not get schema for dataset " + metaClient.getBasePath()));
+          return readSchemaFromDataFile(new Path(filePath));
+        case MERGE_ON_READ:
+          // If this is MOR, depending on whether the latest commit is a delta commit or compaction commit
+          // Get a datafile written and get the schema from that file
+          Optional<HoodieInstant> lastCompactionCommit = metaClient.getActiveTimeline()
+              .getCompactionTimeline().filterCompletedInstants().lastInstant();
+          LOG.info("Found the last compaction commit as " + lastCompactionCommit);
+
+          Optional<HoodieInstant> lastDeltaCommitAfterCompaction = Optional.empty();
+          if (lastCompactionCommit.isPresent()) {
+            lastDeltaCommitAfterCompaction = metaClient.getActiveTimeline()
+                .getDeltaCommitTimeline()
+                .filterCompletedInstants()
+                .findInstantsAfter(lastCompactionCommit.get().getTimestamp(), Integer.MAX_VALUE).lastInstant();
+          }
+          LOG.info("Found the last delta commit after last compaction as "
+              + lastDeltaCommitAfterCompaction);
+
+          if (lastDeltaCommitAfterCompaction.isPresent()) {
+            HoodieInstant lastDeltaCommit = lastDeltaCommitAfterCompaction.get();
+            // read from the log file wrote
+            commitMetadata = HoodieCommitMetadata
+                .fromBytes(activeTimeline.getInstantDetails(lastDeltaCommit).get());
+            filePath = commitMetadata.getFileIdAndFullPaths().values().stream().filter(s -> s.contains(
```

Comment:
Note to self:  we do log the schema with each commit, so we are guaranteed to get a schema from whichever log we read.